### DISCUSSION TITLE
Currently does not work!

### DISCUSSION THREAD

**Discussion by moritzlerch** [moritzlerch/tesla-pv-display#10] [Announcements]
I'm having problems connecting to the powerwall via the ESP8266. Maybe one of you has the time to troubleshoot this, I also did some severe troubleshooting. Thanks in advance! :-)

**Comment by gkostiuk** [comment]
Confirming that it doesn't work. Not competent with this platform, but hopefully a solution or workaround will be found

**Comment by gemu2015** [comment]
hi,
after struggling for a very long time to get the power wall api to run on ESP if finally got it to work.
there is absolutely no chance with the standard ssl libraries included in Arduino.
but the ssl library within [mobitz](https://github.com/mobizt/ESP-Mail-Client) did the trick.
i integrated it into Tasmota, see my fork of tasmota:  https://github.com/gemu2015/Sonoff-Tasmota
if you are interested  i will give further instructions.

<img width="626" alt="Bildschirmfoto 2024-12-04 um 12 17 20" src="https://github.com/user-attachments/assets/e628ca1a-ad35-44d0-8eed-749ca2b6ee7c">

**Comment by gkostiuk** [comment]
Many thanks for your efforts and follow-up,  yes, I am very  interested in further instructions.

Dec. 4, 2024 6:20:25 a.m. gemu ***@***.***>:

**Reply by gemu2015** [reply]
make yourself familiar with tasmota. power wall api only runs on ESP32, so try to flash an ESP32 with Tasmota. first use an existing binary from tasmota to learn how to use it.
if you can't  compile Tasmota yourself i can provide you with a binary containing the api.
you must use Tasmota scripting. i will provide a tasmota script with which you can scan the api for various values and display them on WEBUI or a connected display or send them via MQTT to a broker

**Comment by gkostiuk** [comment]
Thanks for this,  Tasmota has been on my radar for some time. I think I can manage with compilation, but thanks for your offer.

I've used MQTT before to interface my HAI Omni with Hubitat.

Looking forward to your script, much appreciated. Greetings from Toronto.

Dec 4, 2024 10:51:14 a.m. gemu ***@***.***>:

**Reply by gemu2015** [reply]
this is the script.
note: since the tasmota json parser has certain limitations some power wall JSON keywords are renamed to shorter abbreviations by the driver like n_e_r ( = nominal_energy_remaining )

```
>D 100
tmp=0
res=0
cnt=0

; powerwall
pwl=8
;Netz
sip=0
;Solar
sop=0
;Battery
bip=0
;Haus
hip=0
; total cap
tcap=0
; remainig cap
rcap=0
; reserve percent
rper=0

; 3 phases
phs1=0
phs2=0
phs3=0

p1w=0
p2w=0
p3w=0
p4w=0
p5w=0
p6w=0
p7w=0
p8w=0

pwf=0


hs1="{m}<span style='color:"
hs5="</span>"
ps=""
str=""
tm=""
perr=""

>B
; Monate
tmp=is1(0 "Jan|Feb|Mar|Apr|Mai|Jun|Jul|Aug|Sep|Okt|Nov|Dez|")
; Wochentage
tmp=is(0 "So|Mo|Di|Mi|Do|Fr|Sa|")

>BS
; set powerwall ip and credentials (prefix @D)
res=gpwl("@D192.168.188.60,email,password")
; set powerwall serial numbers of CTS devices 1 and 2 (prefix @C)
res=gpwl("@C0x000004714B006CCD,0x000004714B007969")

>S

; call api every 4 seconds 
if year>2023 {
	switch cnt
		case 0
			gpwl("/api/meters/aggregates")
		case 4
			gpwl("/api/system_status/soe")
		case 8
			gpwl("/api/system_status")
		case 12
			gpwl("/api/operation")
		case 14
			pwf=1
			gpwl("/api/meters/readings")
		case 30
			cnt=-1
	ends
	cnt+=1
}

>jp
; fetch powerwall JSON output, some values renamed to shorter labels because tasmota can't handle longer keywords 

sip=site#i_power
bip=battery#i_power
hip=load#i_power
sop=solar#i_power
pwl=percentage
tcap=f_p_e
rcap=n_e_r
rper=b_r_p


; meter readings cannot be decoded by tasmota JSON handler
; so use string search ("gwr") to get the CTS values
if pwf>0
then
	str=""
	str=PW_CTS1#error
	perr=str
	if str==""
	then
		p1w=gwr( "\"p_W\":" 2)
		p2w=gwr( "\"p_W\":" 3)
		p3w=gwr( "\"p_W\":" 4)
		p4w=gwr( "\"p_W\":" 5)
	endif
	str=""
	
	str=PW_CTS2#error
	perr=str
	if str==""
	then
		p5w=gwr( "\"p_W\":" 6)
		p6w=gwr( "\"p_W\":" 7)
		p7w=gwr( "\"p_W\":" 8)
		p8w=gwr( "\"p_W\":" 9)
		phs1=p5w
		phs2=p6w
		phs3=p7w
	endif
	
	pwf=0
	str=""
	
endif


#wtime
; nur in webmode 0 (Tabelle in der auch die Sensoren erscheinen)
if wm==0 {
	; Uhrzeit, Wochentag und Datum anzeigen
	wcs {s}<center><h1 style="color:green;font-size:40px;">%2.0hours%:%2.0mins%:%2.0secs%</h1>{m}%is[wday]% %0day%. %is1[month]% <br>%0year%{e}
	dp(2 0)
	ps=s(int(sunrise/60))+":"+s(sunrise%60)
	str=s(int(sunset/60))+":"+s(sunset%60)
	res=sunset-sunrise
	tm=s(int(res/60))+":"+s(res%60)
	wcs {s}<center>&#127774 %ps% <--- %tm% ---> %str% &#127769{e}
	dp(0 2)
	;wcs {s}<br><a href="/ufs/energy_15_log.txt">15 Minuten Log</a><br>{e}
}

>W
%=#wtime

<hr>
Batterie Füllstand%hs1%yellow;'>%pwl% %% (%2(pwl/100*(tcap/1000))% kWh)%hs5%
Netz%hs1%yellow;'>%sip% W %hs5%
Solar%hs1%green;'>%sop% W %hs5%
Batterie%hs1%yellow;'>%bip% W %hs5%
Haus%hs1%red;'>%hip% W %hs5%
Gesamt Kapazität%hs1%yellow;'>%3(tcap/1000)% kWh %hs5%
Verbleibend%hs1%yellow;'>%3(rcap/1000)% kWh %hs5%
RcapLimit%hs1%yellow;'>%rper% %% %hs5%
<hr>
Solar P1%hs1%yellow;'>%1p1w% W %hs5%
;Netz P2%hs1%yellow;'>%1p2w% W %hs5%
Solar P2%hs1%yellow;'>%1p3w% W %hs5%
;Netz P4%hs1%yellow;'>%1p4w% W %hs5%
<hr>
Netz P1%hs1%yellow;'>%1p5w% W %hs5%
Netz P2%hs1%yellow;'>%1p6w% W %hs5%
Netz P3%hs1%yellow;'>%1p7w% W %hs5%
;Netz P8%hs1%yellow;'>%1p8w% W %hs5%
<hr>
CTS Error%hs1%yellow;'>%perr%%hs5%
<hr>

heap:{m}%heap%

```

**Comment by gkostiuk** [comment]
Once again,  thank you.

Dec 4, 2024 1:15:35 p.m. gemu ***@***.***>:

**Reply by gemu2015** [reply]
the power wall api access has now been merged to Tasmota dev. https://github.com/arendst/Tasmota
i added a script example to the scripting docs that displays several values in the WEBUI and on an ILI9341 LCD display: https://tasmota.github.io/docs/Scripting-Language/#features

thanks again to Moritz Lerch for the initial support of this API on ESP platform

![pw_api](https://github.com/user-attachments/assets/30e17d30-5305-49ea-89b8-7567a0f884a9)
